metaflow/datatools/s3op.py

```diff
@@ -106,66 +157,125 @@ def worker(queue, mode):
+                    s3.upload_file(url.local, url.bucket, url.path, ExtraArgs=extra)
+                    # We indicate that the file was uploaded
+                    result_pipe.send((idx, 0))
 
     except:
         traceback.print_exc()
         sys.exit(ERROR_WORKER_EXCEPTION)
 
 def start_workers(mode, urls, num_workers):
-
+    # We start the minimum of len(urls) or num_workers to avoid starting
+    # workers that will definitely do nothing
+    num_workers = num_workers if len(urls) > num_workers else len(urls)
     queue = Queue(len(urls) + num_workers)
     procs = {}
 
     # 1. push sources and destinations to the queue
-    for url, _ in urls:
-        queue.put(url)
+    idx = 0
+    for elt in urls:
+        queue.put((elt, idx))
+        idx += 1
 
     # 2. push end-of-queue markers
     for i in range(num_workers):
-        queue.put(None)
+        queue.put((None, None))
+
+    # 3. Prepare the result structure
+    sz_results = [None]*idx
 
-    # 3. start processes
+    # 4. start processes
     for i in range(num_workers):
-        p = Process(target=worker, args=(queue, mode))
+        recv, send = Pipe(duplex=False)
+        p = Process(target=worker, args=(send, queue, mode))
         p.start()
-        procs[p] = True
-
-    # 4. wait for the processes to finish
-    while any(procs.values()):
-        for proc, is_alive in procs.items():
-            if is_alive:
-                proc.join(timeout=1)
-                if proc.exitcode is not None:
-                    if proc.exitcode == 0:
-                        procs[proc] = False
-                    else:
-                        msg = 'Worker process failed (exit code %d)'\
-                               % proc.exitcode
-                        exit(msg, proc.exitcode)
+        procs[p] = recv
+        # Do not keep references to the Connection around to make sure
+        # they close properly
+        recv = None
+        send = None
+
+    # 5. wait for the processes to finish; we continuously update procs
+    # to remove all processes that have finished already
+    while procs:
+        new_procs = {}
+        for proc, recv in procs.items():
+            proc.join(timeout=1)
+            if proc.exitcode is not None:
+                # Clean out the final data from the pipe; this is guaranteed
+                # to return EOFError or something immediately
+                try:
+                    while recv.poll():
+                        sz_result = recv.recv()
+                        sz_results[sz_result[0]] = sz_result[1]
+                except EOFError:
+                    pass
+                if proc.exitcode != 0:
+                    msg = 'Worker process failed (exit code %d)'\
+                            % proc.exitcode
+                    exit(msg, proc.exitcode)
+            else:
+                # Do some limited receive to avoid too much buffering on the
+                # pipes.
+                count = 10
```

Comment:
why 10?

Reply:
No particular reason for the number. I did want to set a limit to avoid possible starvation.

Reply:
Addressed in refactoring. This is no longer present.